src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -423,145 +408,154 @@ public void Abort()
         {
-            if (_readingTask == null)
+            if (_readWriteTask == null)
             {
-                _readingTask = ProcessRequestBody();
+                _readWriteTask = ProcessRequestAndResponseBody();
             }
         }
 
-        private async Task ProcessRequestBody()
+        // IIS can only handle on async operation per request at a time.
+        // To guarantee this, we have a single loop to read and write the request/response
+        private async Task ProcessRequestAndResponseBody()
         {
+            var upgrade = false;
+            var doneReading = false;
             try
             {
                 while (true)
                 {
-                    // These buffers are pinned
-                    var wb = Input.Writer.GetMemory(MinAllocBufferSize);
-                    _inputHandle = wb.Retain(true);
-
-                    try
+                    // Read to the input pipe
+                    if (!doneReading)
                     {
-                        int read = 0;
-                        if (_wasUpgraded)
-                        {
-                            read = await ReadWebSocketsAsync(wb.Length);
-                        }
-                        else
-                        {
-                            _currentOperation = _currentOperation.ContinueWith(async (t) =>
-                            {
-                                _currentOperationType = CurrentOperationType.Read;
-                                read = await ReadAsync(wb.Length);
-                            }).Unwrap();
-                            await _currentOperation;
-                        }
-
-                        if (read == 0)
-                        {
-                            break;
-                        }
-
-                        Input.Writer.Advance(read);
+                        doneReading = await ReadToInputPipe();
                     }
-                    finally
+
+                    // Write from the output pipe
+                    if (await WriteFromOutputPipe())
                     {
-                        Input.Writer.Commit();
-                        _inputHandle.Dispose();
+                        break;
                     }
 
-                    var result = await Input.Writer.FlushAsync();
-
-                    if (result.IsCompleted || result.IsCancelled)
+                    if (_upgradeTcs != null)
                     {
-                        break;
+                        upgrade = true;
                     }
                 }
 
-                Input.Writer.Complete();
+                // If we upgraded the stream, we can switch to simultaneously reading and writing
+                // create two new tasks and await them finishing.
+                // We don't want to Complete the Reader and Writer as they will be reused in
+                // the other tasks.
+                if (upgrade && _upgradeTcs.TrySetResult(null))
```

Comment:
Nit: I don't see why you need the upgrade bool. Couldn't you just null check _upgradeTcs here like you do above?